content/how-to-persist-app-data-across-os-updates.md

```diff
@@ -0,0 +1,58 @@
+---
+title: How to Persist Application Data across Torizon OS Updates
+date: 2024-12-02
+author: Leonardo Graboski Veiga, Product Owner Torizon OS
+draft: false
+tags:
+    - Container
+    - Embedded
+    - Linux
+abstract: How to make sure your application data is not overwritten or erased by OS Updates
+image: ./types-of-mounts.webp
+---
+
+The Torizon OS is designed around the concept that different parties handle _applications_ and _OS_ development:
+
+- Applications: you, as a Torizon customer, develop and update your application on your development cycle
+- OS: Toradex, more specifically the Torizon OS Team, develops and updates the base OS, releasing it every quarter
+
+A question I sometimes get from customers is: how can I keep my application data when updating the base OS?
+
+## A Practical Answer
+
+**You must store the application data somewhere under the `/var` directory**. OSTree, the file system management tool used for OS Updates, keeps the `/var` directory intact during the updates.
+
+Therefore, the two standard Docker [storage methods](https://docs.docker.com/engine/storage/) are ok to use.
+
+![Types of Mounts](/types-of-mounts.webp)
+*Figure 1: Types of Mounts (source: Docker Docs)*
+
+You can use [Bind Mounts](https://docs.docker.com/engine/storage/bind-mounts/), as long as keeping the mount somewhere under `/var`. The `/home/torizon` directory is recommended, as it is symlinked to `/var` and owned by the `torizon` user.
+
+Here is an example that shares a directory named `my-app` from the torizon home directory to a container that expects to store data in the directory named `my-app` under `/etc`:
+
+```sh
+docker run -v /home/torizon/my-app:/etc/my-app app-container
+```
+
+You can use [Volumes](https://docs.docker.com/engine/storage/volumes/), as they are effectively stored under /var by default, specifically under /var/lib/docker/volumes. Therefore you can safely use them as documented by Docker.
+
+Here is an example that uses a volume named `my-app` to store data from a container’s directory named `my-app` under `/etc`:
+
+```sh
+docker run -v my-app:/etc/my-app app-container
+```
+
+## Bind Mount or Volume?
+
+The Docker [storage](https://docs.docker.com/engine/storage/) page recommends volumes as the best way to persist data in Docker.
+
+On the one hand, the most compelling use cases for volumes may be irrelevant when considering the Torizon OS use cases: running containers on Windows or having an unknown file system structure.
+
+On the other hand, bind mounts don’t seem to provide any advantage over volumes _for storing application data_.
+
+Therefore it seems natural to choose volumes unless you identify specific needs for a bind mount in your project.
+
+## Conclusion
```

Comment:
I think we could also mention how you can change `/etc` with TCB and use the same techniques shown here to mount the data inside the application containers as well. Then this article could be very "complete".

Reply:
It was a nice catch.

I intentionally avoided `/etc` because I didn't want to increase the complexity by providing an extra option. But after your comment and some contemplation, I see that it would be omitting useful use cases which is worse than the not-so-complex complexity brought with it.